### PR TITLE
Table query

### PR DESCRIPTION
**Pull request by famence**
### Summary:
Adds the ability to table.query() respecting models correctly. Useful for utilising the single table approach when you want to get multiple items of different models within single request.
⚠️ Preferably to be reviewed after https://github.com/dynamoose/dynamoose/pull/1517 (depends on https://github.com/dynamoose/dynamoose/pull/1517) as the current one is based on it.

#### General
Let's assume we have two models Team and TeamMember within single table:
```ts
const Team = dynamoose.model("Team", {
  "pk": {
    "type": String,
    "index": {
      "hashKey": true,
      "index": {
        "name": "primary",
        "type": "global"
      }
    },
    "map": "id"
  },
  "sk": {
    "type": String,
    "rangeKey": true
  },
  "name": String,
  "type": {"type": String, "default": "team"}
});
const TeamMember = dynamoose.model("TeamMember", {
  "pk": {
    "type": String,
    "index": {
      "hashKey": true,
      "index": {
        "name": "primary",
        "type": "global"
      }
    },
    "map": "teamId"
  },
  "sk": {
    "type": String,
    "rangeKey": true,
    "map": "id"
  },
  "name": String,
  "type": {"type": String, "default": "member"}
});
const table = new dynamoose.Table("teams", [Team, TeamMember]);
```
and some instances of each type:
```ts
const team1 = new Team({"id": "team_1", "name": "Team 1"});
const member1 = new TeamMember({"teamId": "team_1", "id": "member_1", "name": "Team member 1"});
```
now we can query the table for any global index (i.e. primary key) and get relevant items with correct models mapping:
```ts
await table.query({"pk": "team_1"}).exec()
// will return the following:
[
  Team {
    id: 'team_1',
    sk: '',
    name: 'Team 1',
    type: 'team'
  },
  TeamMember {
    teamId: 'team_1',
    id: 'member_1',
    name: 'Team member 1',
    type: 'member'
  },
]
```
The models of each item are identified correctly and all the attributes are mapped as we expect in our schemas.

The magic is in respecting attributes' default values which add +2 score weight and in improved Model.schemaCorrectnessScores method by adding up the scores to determine the most suitable model even if the models' schemas are same except default values.

### Type (select 1):
- [ ] Bug fix
- [x] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #---- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/main/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

**Comment by fishcharlie** [comment]
@famence #1517 has been merged. Do you maybe want to merge `develop` into this branch and finalize anything else and mark it ready for review and I can take a look at it?

**Comment by famence** [comment]
@fishcharlie i just finished the technical part so you are welcome to review the code

**Comment by famence** [comment]
Documentation updated

**Comment by marcofailache** [comment]
Hey, guys! Thanks for the hard work!
Any timeframe for when this pull request will be merged?

File: packages/dynamoose/lib/Table/index.ts [hunk 1]
```diff
@@ -205,10 +207,10 @@ export class Table extends InternalPropertiesClass<TableInternalProperties> {
 					"arrayItemsMerger": utils.merge_objects.schemaAttributesMerger
 				})(...createTableAttributeParams);
 			},
-			"getHashKey": (): string => {
+			"getHashKey": () => {
```
Comment (fishcharlie) [inline packages/dynamoose/lib/Table/index.ts:210]: Any reason why we are removing these return types?

Comment (famence) [reply packages/dynamoose/lib/Table/index.ts]: Brought it back @https://github.com/dynamoose/dynamoose/pull/1538/commits/2e5cb52d97f69269dc961e9b9bf2b7a04c804a8d

**Comment by saltman424** [comment]
I know it's been a little over a year, but just wanted to check: is there still potential for this to be merged?

**Comment by fishcharlie** [comment]
@saltman424 This fell off my radar. Do me a favor and try to help remind me this weekend and I'll take a look. This week is pretty busy for me so probably won't get around to it until this weekend.

**Comment by saltman424** [comment]
@fishcharlie sorry, I meant to ping you earlier in the weekend, but consider this the requested reminder

**Review by fishcharlie — APPROVED** [review]

**Comment by fishcharlie** [comment]
This PR has been reverted in Dynamoose version 4.0.3. See #1702 & #1714 for more information.

**Pull request merged**